### ISSUE TITLE
V4 Should give refund for an overbuy

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#5194]
**Describe the bug**
TODO: find a good example with V4 rules

In v4 for example a refund for an overbuy should be given, the overbuys are lost without refund instead.

In Global 1940 a refund for an unintentional overbuy should be given, the overbuys are carried over to the next turn(s) instead.

**To Reproduce**
- start game
- purchase more units than can be placed
- end turn phase
- notice: (?? TODO: what is the current behavior on v4 maps?)

**Expected behavior**
- cash refund given to the player
- unplaced purchased units removed

**Comment by Cernelius** [comment]
The refund is to be actually provided only for unintentional overbuy. It is thus propedeutic to this feature to define what makes an overbuy unintentional. Also, there are no rules to define what happens in case of intentional overbuy, so I would argue that the outcome would be that you lose the game, if the overbuy is somehow proved intentional. Thus also a feature making you lose the game if you do an intentional overbuy would have to be provided, as well.

**Comment by panther2** [comment]
@Cernelius Indeed. We have discussed the vagueness of that "unintentional overbuy" in other related Github issues as far as I remember. When I find some time I will reference them.

**Comment by Cernelius** [comment]
Yeah, it must be exactly the same as v5 and Global. I'm fairly sure Krieghund will say that.

**Comment by panther2** [comment]
@Cernelius @DanVanAtta 

... and he did confirm that.
https://www.axisandallies.org/forums/post/1332805

So the rule for v4 is "Overbuy is forbidden, there is no refund (except "unintentional").", the same as for v5, v6 and Global, Europe, Pacific 1940 2nd. Edition (https://github.com/triplea-game/triplea/issues/5195).

**Comment by panther2** [comment]
The headline of this issue should be correted, IMHO.

**Comment by Cernelius** [comment]
@panther2 Correct, I suggest changing the title from:

"overbuy should provide a refund"

to, either:

"**overbuy should make you lose the game, if intentional, otherwise provide a refund**"

or

"**overbuy should be impossible**".

In the first case, you should determine a way to distinguish between intentional and unintentional overbuys. In the second case, TripleA should never allow you overbuy.

However, in the second case, you would also have the challenge of supporting placement relevant units that can be moved, in any phases in between, and players having the ability of placing in newly conquered territories.

I also suggest expanding this issue as to encompass all current basic games from v4 onwards.
https://github.com/triplea-game/triplea/wiki/Game-Rule-Sets

Finally, @panther2 I would argue that this rule exists since v1 (Classic). While games deal with the matter differently (all units lost or kept in inventory or refunded), I believe all of them assume the overbuy must have been unintentional, are they not? If so, there would be absolutely no difference for any basic games, from Classic to the latest, if TripleA would enforce the no overbuy limit flawlessly itself.

**Comment by panther2** [comment]
@Cernelius In principle yes, I would agree. However in v1 I recall that you are not allowed to mobilize ships in hostile seazones. So if you bought a ship that you intended to place in an enemy controlled seazone adjacent to your IC and the attack fails, you will not be allowed to place the ship in that seazone. Provided there is no other eligible seazone to place, I would not call this an "intentional overbuy".

Also an intentional overbuy should not contain facilities in general, as those don't count against the production limits. 

And, last not least, issues such as #3506 are independent from the topic discussed here.

**Comment by Cernelius** [comment]
>Also an intentional overbuy should not contain facilities in general, as those don't count against the production limits.

I assume you mean that all the different placement cases should be accounted correctly, for overbuy. I believe overbuy matters also for facilities. For example, if all my territories have a factories, I cannot buy any factories, can't I?

And, yes, of course, overbuy should be determined based on everything having at least a chance to be placed (which, I would add, should be calculated under regular dice rules). Substantially the same principle of legal air movements. In a game like Napoleonic Empires, you may have a territory you can place into by offloading a general into it that is possible only as long as you successfully clear a sea zone, for example.

**Comment by DanVanAtta** [comment]
Is this a fix that is needed in game engine, or is this actually a fix that is needed in map XML?

@Cernelius / @ron-murhammer / @panther2 , any insight?

**Comment by panther2** [comment]
While I actually don't know what of this is engine based and what is XML-based I think we first need to dicuss/decide on how to move on based on the above discussion.

**Comment by Cernelius** [comment]
It surely cannot be fixed in the xml alone. After fixing it in the engine, and depending on how this is fixed, updating every game xml may be virtually necessary, because of the usual issue that, after "v2", no real general properties were anymore given and, after "v3", no general properties at all. So, there is no way, for example, of telling apart v3 and v4 rules games, for example. If v3 and v4 rules games would have to work differently, depending on how the engine is fixed, then I guess you would need to search for all of the 100+ games in the repository and update their xml accordingly, if you can even know or guess what ruleset they are following.

**Comment by DanVanAtta** [comment]
>  and depending on how this is fixed

We need to define that here, work can't start until we do.
AFAIK the only thing needed in the OP is just an example map for it to be more clear what the requirement is. I think we are making this issue more generic than it needs to be. By expanding the scope and complexity of the fix, we're in effect tying additional problems together that make it less likely that any of the problems are going to be fixed.

**Comment by Cernelius** [comment]
I see only a generally workable solution (all these points are needed):

- The engine is made infallible in determining what you can purchase, so that you are never allowed to purchase anything, unless you have at least a chance to place it, substituting the current confirmation prompt on overpurchase with a strict prohibition to end the phase in such a condition (**an opt out property, to allow overbuy should be still provided, to support custom maps that want to allow deliberately overpurchasing or that have features so complex that the engine won't cover**).
- Once in the placement phase, you cannot end the phase if you can still place anything, on top of what you already placed.
- Once the placement phase ends, all unplaced units are removed and their value in any kind of resources is added to the owner.

**Comment by panther2** [comment]
1940 first edition has the v4 behaviour.

**Comment by Cernelius** [comment]
> I see only a generally workable solution (all these points are needed):
> 
> 1- The engine is made infallible in determining what you can purchase, so that you are never allowed to purchase anything, unless you have at least a chance to place it, substituting the current confirmation prompt on overpurchase with a strict prohibition to end the phase in such a condition (**an opt out property, to allow overbuy should be still provided, to support custom maps that want to allow deliberately overpurchasing or that have features so complex that the engine won't cover**).
> 2- Once in the placement phase, you cannot end the phase if you can still place anything, on top of what you already placed.
> 3- Once the placement phase ends, all unplaced units are removed and their value in any kind of resources is added to the owner.
> 
> If the first and second points work, the third point will actually never happen, in the v4 game.
> 
> I believe:
> 
> * Point 1 should apply to all rulesets, from v1 to v6.
> * Point 2 should apply to all rulesets, from v1 to v6.
> * Point 3 should apply to v4 and all following rulesets (comprising E/P/G 1940 II).
> 
> The only thing that I don't remember is if E/P/G 1940 I (like Global 1940 1st Edition) has the v3 or v4 behaviour for unplaced, and how about all its Alpha.

I guessed wrong on point 1 and 2, so here are them all reformulated and, now, hopefully correct:

---

I see only a generally workable solution, for v4 and following rules sets (all these points are needed):

1- The engine is made infallible in determining what you can purchase, so that you are never allowed to purchase anything, unless you have at least a chance to place all units you are purchasing, substituting the current confirmation prompt on overpurchase with a strict prohibition to end the phase in such a condition (in practice, the program would take over the role of your opponent in checking you are not pulling any overpurchase).
2- Once in the placement phase, you cannot end the phase if you have a number of units still not placed on the gameboard, outside the mobilization zone, that is greater than the minimum number of them you could have possibly left unplaced, starting from the state of the game at the start of the current phase.
3- Once the placement phase ends, all unplaced units are removed and their value in any kind of resources is added to the owner.

If the first point works, the second and third points will actually never happen, in the v4 game (but can happen in custom games based on v4 rules, for example by having mobile factories).

Moreover:

- Point 1 should apply to all rules sets from v4 to v6 (comprising E/P/G 1940 I and II, comprising every "alpha"). Instead, all rules sets from v1 to v3 should allow you to overpurchase, even intentionally (providing an "are you sure" warning would be still opportune).
- Point 2 should apply to all rules sets, from v1 to v6.
- Point 3 should apply to all rules sets from v4 to v6 (comprising E/P/G 1940 I and II, comprising every "alpha"). Instead, the v3 rules set should keep your non-placed units inside your inventory, so that you can place them on any following placement phase. Instead, all rules sets from v1 to v2 (comprising the original (non-1940) Europe and Pacific) should not add any resources to the owner (units are removed from the game with no refund).

Regarding point 1, an opt out property, to allow overbuy should be still provided, to support custom maps that want to allow deliberately overpurchasing or that have features so complex that the engine won't cover, while otherwise possibly following mostly or totally v4 or following rules (for example, one might make a (likely more complex) modification of "WWII Global", in which it is needed to be able intentionally to overpurchase, even though in the original this should never be possible).

Regarding point 1 and 2, I advise being able to end any phase if "Edit" mode is enabled. This would be consistent with "Edit" mode generally allowing you to violate the rules.

@panther2 Please confirm I got everything right, this time.
@DanVanAtta If confirmed, please remove the "Needs Clarification" label or clarify what still needs clarification.

---

@panther2 Off topic, but is what at this post the only general rules difference between v3 and v4 or are there others? If it is, can we consider v4 more like as a fixed version of v3, where you are tentatively disabled from abuse the purchase and placement rules, than an actual "new" rules set? If so, why not retroactively apply such a small change to v3 too (when remaking the rulebook), as an errata, to avoid upkeeping such a minor differentiation?

**Comment by panther2** [comment]
@Cernelius Indeed, I am not aware of any other rules difference between v3 and v4. I have always considered v4 as being a simplification of v3 (as v5 is of Global 1940 2nd ed.), and definitely not as being a different ruleset (apart from the "overbuy-rule"). 
I would not object if that change was implemented retroactively into v3.

**Comment by panther2** [comment]
@Cernelius I am not sure that I have understood point 2 correctly.
Let's assume that at the beginning of the turn you are able to place 10 units during placement phase.
Now you are overbuying and place 12 units in the mobilization zone.
Does point 2 say that the placement phase may not end until you have placed 10 units, thus prohibiting not to place 3 or more units?

**Comment by Cernelius** [comment]
Point 2 means you must place the maximum number of units (no matter their cost) you can possibly place. I just worded this from what I think it should be the point of view of the engine, giving you the ability to end the phase only if it doesn't see a way to have less units unplaced (thus to refund, if v4+). The difference from my previous formulation is that you can reach a point where you can place no more units, but this is not a situation in which you can legally end the phase if, with different placement actions, you could have managed to place a higher number of units from your inventory.

**Comment by panther2** [comment]
Got it, good point!
Your summary looks good!

**Comment by panther2** [comment]
> @DanVanAtta If confirmed, please remove the "Needs Clarification" label or clarify what still needs clarification.

Just bringing up and joining @Cernelius question...
